src/lib.rs

```diff
@@ -63,9 +63,18 @@
 //! ```
 
 use proc_macro::TokenStream;
-use quote::quote;
+use quote::{format_ident, quote};
 use syn::{parse_macro_input, parse_quote};
 
+macro_rules! build_seal {
+    ($seal:ident) => {{
+        // TODO there has to be a better way
+        use heck::SnakeCase as __heck__SnakeCase;
+        let seal_ident = $seal.to_string().to_snake_case();
+        format_ident!("__seal_for_{}", seal_ident)
+    }};
+}
```

Comment:
Certainly, it is a macro because it started as a what to have a single place for `__seal__for_{}` and a macro which expands to a literal was the best choice I could come up with

Reply:
Also, why `fn seal_name(seal: &'_ dyn ::core::fmt::Display) -> ::syn::Ident` instead of `fn seal_name<D: Display>(seal: D) -> syn::Ident`?

Reply:
As to why use `dyn`amic dispatch over static dispatch: I've been trying to _default_ to `dyn`amic dispatch in places where it will clearly make no much difference in terms of it being a super tiny bit slower, since every time static dispatch is used instead, it does lead to a slightly slower compilation time (and slightly more bloated binary, although that doesn't apply to proc-macros). Here, in particular, the time "saved" running the proc-macro _at compile time_ using static dispatch is very likely to be trumped by the time spent on compiling the static dispatch. My TL,DR in that regard is to favor `dyn` dispatch when:

  - dealing with proc-macro code,

  - dealing with a trait performing clearly more expensive (that virtual method calls) operations (such as `io::{Read, Write}`).

For context, an imho must read blog post: https://pingcap.medium.com/the-rust-compilation-model-calamity-1a8ce781cf6c